servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpConnector.java

```diff
@@ -140,13 +125,43 @@ protected void handleSubscribe(final Subscriber<? super C> subscriber) {
     private static Future<?> connect0(@Nullable SocketAddress localAddress, Object resolvedRemoteAddress,
                                       ReadOnlyTcpClientConfig config, boolean autoRead,
                                       ExecutionContext<?> executionContext,
-                                      Consumer<? super Channel> subscriber) {
+                                      ConnectHandler<?> connectHandler) {
         // Create the handler here and ensure in connectWithBootstrap / initFileDescriptorBasedChannel it is added
         // to the ChannelPipeline after registration is complete as otherwise we may miss channelActive events.
         ChannelHandler handler = new io.netty.channel.ChannelInitializer<Channel>() {
             @Override
             protected void initChannel(Channel channel) {
-                subscriber.accept(channel);
+                connectHandler.accept(channel);
+                // We need to intercept the `connect` call here and add our listener because right after the pipeline
+                // gets its `connect(..)` sequence called the netty Bootstrap will add the `CLOSE_ON_FAILURE` listener
+                // to the connect future. That will bypass the `assignConnectionError` and cause us to miss the reason
+                // for the channel closing.
+                channel.pipeline().addLast(new ChannelOutboundHandlerAdapter() {
+                    @Override
+                    public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress,
```

Comment:
This is a little ugly but I don't know of a better way to intercept the `ChannelPromise` associated with the connect operation and register our own callbacks before the event loop can register it's `CLOSE_ON_FAILURE` listener [here](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/bootstrap/Bootstrap.java#L268). Suggestions welcome.

Reply:
Good find!
I don't see another way to intercept it earlier. The only suggestion is to move addition of this handler before `connectHandler.accept(channel);` to make sure nothing can throw earlier.